src/core/engines/BaseEngine.sol

```diff
@@ -25,21 +27,75 @@ import "../../config/errors.sol";
  * @author  @antoncoding
  * @notice  util functions for MarginEngines
  */
-contract BaseEngine {
+abstract contract BaseEngine is ReentrancyGuard {
     using SafeERC20 for IERC20;
     using TokenIdUtil for uint256;
 
+    IGrappa public immutable grappa;
+    IOptionToken public immutable optionToken;
+
     ///@dev maskedAccount => operator => authorized
     ///     every account can authorize any amount of addresses to modify all sub-accounts he controls.
     mapping(uint160 => mapping(address => bool)) public authorized;
 
     /// Events
     event AccountAuthorizationUpdate(uint160 maskId, address account, bool isAuth);
 
+    event CollateralAdded(address subAccount, address collateral, uint256 amount);
+
+    event CollateralRemoved(address subAccount, address collateral, uint256 amount);
+
+    event OptionTokenMinted(address subAccount, uint256 tokenId, uint256 amount);
+
+    event OptionTokenBurned(address subAccount, uint256 tokenId, uint256 amount);
+
+    event OptionTokenMerged(address subAccount, uint256 longToken, uint256 shortToken, uint64 amount);
+
+    event OptionTokenSplit(address subAccount, uint256 spreadId, uint64 amount);
+
+    event OptionTokenAdded(address subAccount, uint256 tokenId, uint64 amount);
+
+    event OptionTokenRemoved(address subAccount, uint256 tokenId, uint64 amount);
+
+    event AccountSettled(address subAccount, uint256 payout);
+
+    /** ========================================================= **
+                            External Functions
+     ** ========================================================= **/
+
+    constructor(address _grappa, address _optionToken) {
+        grappa = IGrappa(_grappa);
+        optionToken = IOptionToken(_optionToken);
+    }
+
     /** ========================================================= **
-                            Internal Functions
+                            External Functions
      ** ========================================================= **/
 
+    function execute(address _subAccount, ActionArgs[] calldata actions) public virtual nonReentrant {
```

Comment:
@antoncoding

Reply:
we can let them revert early by putting `execute` into each engine. (if action not matched, revert), instead of waiting until each internal function implementation to revert.

Trade off is if we put the execute into each engine, we will have quite some duplications